### PR TITLE
get rid of gms/failure_detector

### PR DESCRIPTION
**Pull request by bhalevy**
Move gms::arrival_window to api/failure_detector which is its only user.
and get rid of the rest, which is not used, now that we use direct_failure_detector instead.

TODO: integare direct_failure_detector with failure_detector api.

### CODE REVIEW

File: test/rest_api/test_failure_detector.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import re
+
+def test_failure_detector_endpoints(rest_api):
+    resp = rest_api.send('GET', "failure_detector/endpoints")
+    resp.raise_for_status()
+    assert len(resp.json()) == 1
+    addr = resp.json()[0]['addrs']
+    re.match(r'\d+\.\d+\.\d+\.\d+', addr)
```
Comment (tgrabiec) [inline test/rest_api/test_failure_detector.py:12]: Result of match() is ignored, so the statement has no effect.

File: api/failure_detector.cc [hunk 2]
```diff
@@ -8,9 +8,94 @@
...
 #include "gms/application_state.hh"
 #include "gms/gossiper.hh"
+#include "utils/bounded_stats_deque.hh"
+#include "log.hh"
+
+extern logging::logger apilog;
+
+namespace gms {
+
+class arrival_window {
```
Comment (tgrabiec) [inline api/failure_detector.cc:20]: The only use is here:

```

    fd::get_endpoint_phi_values.set(r, [](std::unique_ptr<request> req) {
        std::map<gms::inet_address, gms::arrival_window> map;
        std::vector<fd::endpoint_phi_value> res;
        auto now = gms::arrival_window::clk::now();
        for (auto& p : map) {
            fd::endpoint_phi_value val;
            val.endpoint = p.first.to_sstring();
            val.phi = p.second.phi(now);
            res.emplace_back(std::move(val));
        }
        return make_ready_future<json::json_return_type>(res);
    });
```

Notice that `map` is not initialized, so it's empty. Before 0ef33b71bab2b15f8f42acea44fd4049fafe46e2, it was obtained from the gossiper gms::get_arrival_samples().

So you can just remove arrival_window altogether.

I don't think the logic of arrival_window belongs to api, it belongs to the failure detector. If there is no longer a failure detector, there should be no arrival_window.

**Review by tgrabiec — CHANGES_REQUESTED** [review]

**Comment by bhalevy** [comment]
In v2 ([cfe69f5](https://github.com/scylladb/scylladb/commit/cfe69f59e52beaa9b3200ee7cc0a48ea13950099)):
- fixed unit test to assert the value of re.match
- removed degenerate use of failure_dector::arrival_window in api
- got rid of gms/failure_detector.{cc,hh} as a whole

File: test/rest_api/test_failure_detector.py (hunk 1 above)
Comment (bhalevy) [reply test/rest_api/test_failure_detector.py]: fixed in v2

File: api/failure_detector.cc (hunk 2 above)
Comment (bhalevy) [reply api/failure_detector.cc]: fixed in v2

**Review by tgrabiec — APPROVED** [review]

File: api/failure_detector.cc [hunk 3]
```diff
@@ -81,15 +80,10 @@ void set_failure_detector(http_context& ctx, routes& r, gms::gossiper& g) {
     });
 
     fd::get_endpoint_phi_values.set(r, [](std::unique_ptr<request> req) {
-        std::map<gms::inet_address, gms::arrival_window> map;
+        // FIXME: this was based on gms::get_arrival_samples.
+        // But the latter returned an empty std::map<inet_address, arrival_window>.
+        // The functionality needs to be implemented by the failure detector.
```
Comment (kbr-scylla) [inline api/failure_detector.cc:85]: We no longer have a phi failure detector, so the functionality will never be implemented.
Maybe we should just remove the API. Even if someone depends on it (I doubt it) it wasn't returning anything meaningful for some time already.

**Review by kbr-scylla — APPROVED** [review]

**Comment by avikivity** [comment]
@asias please review too

File: api/failure_detector.cc (hunk 3 above)
Comment (asias) [reply api/failure_detector.cc:85]: We do not need the FIXME, just returning the empty value is good enough.  The phi values associated with the old fd is gone and is not going back.

**Comment by asias** [comment]
Looks good. I have a minor comment inline.

**Comment by bhalevy** [comment]
In v3 ([3f1ac84](https://github.com/scylladb/scylladb/commit/3f1ac846d85fd1df12f0e631e11d504cc29e0a7a)):
- rebased
- turned FIXME comment into a regular one
  - just explaining why the api function returns an empty value

**Review by asias — APPROVED** [review]

**Comment by kbr-scylla** [comment]
Qd

**Pull request merged**